src/Castle.ts

```diff
@@ -0,0 +1,198 @@
+import fetch from 'node-fetch';
+import { reduce } from 'lodash';
+import { IncomingHttpHeaders } from 'http2';
+
+const apiUrl = 'https://api.castle.com';
+
+type TrackParameters = {
+  event: string;
+  user_id: string;
+  user_traits: any;
+  created_at?: string;
+  context: {
+    ip: string;
+    client_id: string;
+    headers: IncomingHttpHeaders;
+  };
+};
+
+type AuthenticateResult = {
+  action: string;
+  user_id?: string;
+  device_token?: string;
+  failover?: boolean;
+  failover_reason?: string;
+};
+
+export class Castle {
+  private apiSecret: string;
+  private timeout: number;
+  private allowedHeaders: string[];
+  private disallowedHeaders: string[];
+  private overrideFetch: any;
+  private failoverStrategy: string;
+
+  constructor({
+    apiSecret,
+    timeout = 1000,
+    allowedHeaders = [],
+    disallowedHeaders = [],
+    overrideFetch = fetch,
+    failoverStrategy = 'allow',
+  }: {
+    apiSecret: string;
+    timeout?: number;
+    allowedHeaders?: string[];
+    disallowedHeaders?: string[];
+    overrideFetch?: any;
+    failoverStrategy?: string;
+  }) {
+    if (!apiSecret) {
+      throw new Error(
+        'Castle: Unable to instantiate Castle client, API secret is missing.'
+      );
+    }
+
+    this.apiSecret = apiSecret;
+    this.timeout = timeout;
+    this.allowedHeaders = allowedHeaders;
+    this.disallowedHeaders = disallowedHeaders;
+    this.overrideFetch = overrideFetch;
+    this.failoverStrategy = failoverStrategy;
+  }
+
+  public async authenticate(
+    params: TrackParameters
+  ): Promise<AuthenticateResult> {
+    if (!params.event) {
+      throw new Error('Castle: event is required when calling authenticate.');
+    }
+
+    let response: Response;
+
+    try {
+      response = await this.getFetch()(`${apiUrl}/v1/authenticate`, {
+        method: 'POST',
+        timeout: this.timeout,
+        headers: this.generateDefaultRequestHeaders(),
+        body: this.generateRequestBody(params),
+      });
+    } catch (e) {
+      if (e.message.startsWith('network timeout')) {
```

Comment:
just a question - isn't there a better way to detect the timeout than the error message? I find that method in ruby really unreliable (but maybe this is how it goes here ;) )